### PR TITLE
feat: added to DRG module the option to create a Remote Peering Connection (RPC) with a remote VCN

### LINKED ISSUE #71: Add ability to create RPC
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
Add ability to create a Remote Peering Connection (RPC) with another VCN. Best if we can add it in a submodule inside VCN instead of  top-level.

### Potential Terraform Configuration

```hcl
# Copy-paste any Terraform configurations for how the requested feature may be used.
```

### PR DESCRIPTION
**Pull request by snafuz**
close #71 

# Proposed change
Add remote VCN peering capability, enabling Remote Peering Connection into the DRG module.

2 examples are provided to show how to use the new feature both using drg module and vcn module (for backward compatibility) 

Signed-off-by: Andrea Marchesini <andrea.marchesini@oracle.com>

## How has these changes been tested?
Both examples have been tested, verifying that the peering connection is up and running and traffic can be exchanged between instances deployed in local and remote regions.

## Checklist before submitting your PR

- [x ] My code follows [the style guidelines of this project](../tree/main/docs/codingconventions.adoc)
- [ x] these changes provision new resources:
  - [x] I have updated the README introduction section (README.adoc)
  - [x] I have updated the README introduction section (README.md)
- [x] these changes adds any new variables:
  - [x] I have updated [docs/terraformoptions.adoc](../tree/main/docs/terraformoptions.adoc) to include each variable
- [x] I have updated the changelog to include an entry for these changes
- [x] I have updated all provided examples, including each README file and all applicable code blocks
- [x] these changes generates no new warnings
- [x] Any dependent changes have been merged and published in upstream modules

Note: *If you are not an Oracle employee, to contribute to an Oracle-sponsored open-source project, you need to sign the [Oracle Contributor Agreement (OCA)](https://oca.opensource.oracle.com/).*

### CODE REVIEW

File: examples/rpc/rpc_from_vcn_module/main.tf
Comment (kral2) [inline examples/rpc/rpc_from_vcn_module/main.tf:20]: Better to use the terraform public registry path in examples so the example is ready to run without requirement to clone the git repo first

```suggestion
  source  = "oracle-terraform-modules/vcn/oci"
```

File: examples/rpc/rpc_from_vcn_module/main.tf
Comment (kral2) [inline examples/rpc/rpc_from_vcn_module/main.tf:81]: Better to use the terraform public registry path in examples so the example is ready to run without requirement to clone the git repo first

```suggestion
  source  = "oracle-terraform-modules/vcn/oci"
```

File: examples/rpc/rpc_from_vcn_module/main.tf
Comment (kral2) [inline examples/rpc/rpc_from_vcn_module/main.tf:115]: Please use resources instead of external modules we don't maintain directly. This is to reduce external dependencies.
```suggestion
resource "oci_core_subnet" "remote" {
```

File: examples/rpc/rpc_from_vcn_module/main.tf
Comment (kral2) [inline examples/rpc/rpc_from_vcn_module/main.tf:51]: Please use resources instead of external modules we don't maintain directly. This is to reduce external dependencies.
```suggestion
resource "oci_core_subnet" "local" {
```

**Review by kral2 — CHANGES_REQUESTED** [review]
Thanks for your contribution Andrea! RPC support is an important feat we need to get done :-)

I just review some of the example files, and there is some points that would need to be addressed before merging.

Will need another round for the module's logic itself.

File: examples/rpc/rpc_from_drg_module/main.tf
Comment (hyder) [inline examples/rpc/rpc_from_drg_module/main.tf:61]: Please format

File: examples/rpc/rpc_from_drg_module/main.tf
Comment (hyder) [inline examples/rpc/rpc_from_drg_module/main.tf:67]: There's a double // here

**Review by hyder — CHANGES_REQUESTED** [review]
For the examples, can you please also add a diagram to illustrate?

File: examples/rpc/rpc_from_drg_module/main.tf
Comment (kral2) [reply examples/rpc/rpc_from_drg_module/main.tf:67]: That's not a typo. It is the correct syntax to access a submodule directly from the registry.

Comment (hyder) [reply examples/rpc/rpc_from_drg_module/main.tf:67]: Ok.

**Review by hyder — CHANGES_REQUESTED** [review]
A few more things waiting to be resolved. I think for RPC, functionally it's good. But I want to make sure we get the naming right. `local`/`remote` doesn't cut it for me.

We need to name them properly, maybe something like this:

label-region[-optional_name]

Usually, it should be only label-region but if a user happens to create to VCNs in the same region, then they need to be able to distinguish the RPCs.

Once we settle on the naming of this, we also need to make sure it's in our docs and the images as well.

File: examples/rpc/rpc_from_drg_module/README.md
Comment (snafuz) [reply examples/rpc/rpc_from_drg_module/README.md:14]: I've refactored the code adopting requestor/acceptor approach

File: examples/rpc/rpc_from_drg_module/README.md
Comment (hyder) [inline examples/rpc/rpc_from_drg_module/README.md:61]: The alias should also be `requestor`

File: examples/rpc/rpc_from_drg_module/main.tf
Comment (hyder) [inline examples/rpc/rpc_from_drg_module/main.tf:21]: The version of the module is also required

File: examples/rpc/rpc_from_drg_module/main.tf
Comment (hyder) [inline examples/rpc/rpc_from_drg_module/main.tf:91]: Version of the module is required

File: examples/rpc/rpc_from_drg_module/terraform.tfvars.example
Comment (hyder) [inline examples/rpc/rpc_from_drg_module/terraform.tfvars.example:33]: VCN cidr is a list of strings

File: examples/rpc/rpc_from_vcn_module/README.md
Comment (hyder) [inline examples/rpc/rpc_from_vcn_module/README.md:61]: alias should also be `requestor`

File: examples/rpc/rpc_from_vcn_module/terraform.tfvars.example
Comment (hyder) [inline examples/rpc/rpc_from_vcn_module/terraform.tfvars.example:29]: vcn cidr block is a list of strings

File: examples/rpc/rpc_from_vcn_module/terraform.tfvars.example
Comment (hyder) [inline examples/rpc/rpc_from_vcn_module/terraform.tfvars.example:33]: vcn cidr block is a list of strings

File: examples/rpc/rpc_from_vcn_module/variables.tf
Comment (hyder) [inline examples/rpc/rpc_from_vcn_module/variables.tf:73]: vcn cidr block is a list of strings

File: examples/rpc/rpc_from_vcn_module/variables.tf
Comment (hyder) [inline examples/rpc/rpc_from_vcn_module/variables.tf:74]: as above

File: examples/rpc/rpc_from_vcn_module/variables.tf
Comment (hyder) [inline examples/rpc/rpc_from_vcn_module/variables.tf:81]: vcn cidr block is a list of strings

**Review by hyder — CHANGES_REQUESTED** [review]
some naming changes required. Also, VCN cidr blocks takes an updatable list of strings

File: examples/rpc/rpc_from_drg_module/README.md
Comment (hyder) [inline examples/rpc/rpc_from_drg_module/README.md:12]: Use backticks only for code

File: examples/rpc/rpc_from_drg_module/main.tf
Comment (hyder) [inline examples/rpc/rpc_from_drg_module/main.tf:124]: We route _through_ an RPC to another VCN but _not_ to an RPC.

Comment (hyder) [reply examples/rpc/rpc_from_drg_module/main.tf:124]: Please amend description

File: examples/rpc/rpc_from_vcn_module/main.tf
Comment (hyder) [inline examples/rpc/rpc_from_vcn_module/main.tf:99]: As above. I think just call it acceptor VCN

File: examples/rpc/rpc_from_drg_module/main.tf
Comment (hyder) [reply examples/rpc/rpc_from_drg_module/main.tf:124]: I think just call it acceptor VCN

**Review by hyder — CHANGES_REQUESTED** [review]
Minor cosmetic changes

**Review by hyder — APPROVED** [review]
Approved

File: examples/rpc/rpc_from_drg_module/main.tf
Comment (kral2) [reply examples/rpc/rpc_from_drg_module/main.tf:22]: The more I think about pinning version in modules examples or not, the more I tend to say no. Here is the rational behind my current view on that matter.

The problem is that the example's files are not released yet, hence version 3.2.0 will not be able to support this example: user will get an error when trying to copy/paste and execute the example.

In addition, we would always need to come back to example files and keep module version current.

Letting modules in examples without versions on purpose seems to be a better tradeoff:
- less maintenance,
- released versions on the registry (that is, with a tag) always works.
- bleeding edge users can always source the module using the git pattern url and use repo's `HEAD` (great for testing by the way).

But let's settle this later and not block the PR from being merged into `main`. If we choose to not pin module versions in examples, we will to search and update every occurrences of it before cutting the next release tag.

**Review by kral2 — COMMENTED** [review]
Looks good to me, after addressing the two wording issue.

**Review by kral2 — APPROVED** [review]

**Pull request merged**
